src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.extra.ffi;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import jnr.ffi.Runtime;
+import org.truffleruby.core.FinalizationService;
+import sun.misc.Unsafe;
+
+import java.lang.reflect.Field;
+
+public class Pointer implements AutoCloseable {
+
+    public static final jnr.ffi.Pointer JNR_NULL = Runtime.getSystemRuntime().getMemoryManager().newOpaquePointer(0);
+
+    public static final Pointer NULL = new Pointer(0);
+
+    public static Pointer malloc(long size) {
+        return new Pointer(UNSAFE.allocateMemory(size));
+    }
+
+    public static Pointer calloc(long size) {
+        final Pointer pointer = malloc(size);
+        pointer.writeBytes(0, size, (byte) 0);
+        return pointer;
+    }
+
+    private final long address;
+    private boolean autorelease;
+
+    public Pointer(long address) {
+        this.address = address;
+    }
+
+    public void writeByte(long offset, byte b) {
+        UNSAFE.putByte(address + offset, b);
+    }
+
+    public void writeShort(long offset, short value) {
+        UNSAFE.putShort(address + offset, value);
+    }
+
+    public void writeInt(long offset, int value) {
+        UNSAFE.putInt(address + offset, value);
+    }
+
+    public void writeLong(long offset, long value) {
+        UNSAFE.putLong(address + offset, value);
+    }
+
+    public void writePointer(long offset, Pointer value) {
+        writeLong(offset, value.getAddress());
+    }
+
+    public void writeZeroTerminatedBytes(long offset, byte[] bytes, int start, int length) {
+        writeBytes(offset, bytes, start, length);
+        writeByte(offset + length, (byte) 0);
+    }
+
+    public void writeBytes(long offset, long size, byte value) {
```

Comment:
There is `Unsafe.setMemory()` for this use-case, which is likely more efficient for this.